File: documentapi/src/main/java/com/yahoo/documentapi/messagebus/protocol/LoadBalancerPolicy.java
// Copyright 2017 Yahoo Holdings. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root.
package com.yahoo.documentapi.messagebus.protocol;

import com.yahoo.jrt.slobrok.api.Mirror;
import com.yahoo.messagebus.ErrorCode;
import com.yahoo.messagebus.Reply;
import com.yahoo.messagebus.routing.Hop;
import com.yahoo.messagebus.routing.Route;
import com.yahoo.messagebus.routing.RoutingContext;
import com.yahoo.messagebus.routing.RoutingNodeIterator;

import java.util.List;
import java.util.Map;

/**
 * Routing policy to load balance between nodes in a randomly distributed cluster, such as a docproc cluster.
 *
 * pattern=&lt;pattern&gt; (mandatory, determines the pattern of nodes to send to)<br>
 * slobroks=&lt;comma-separated connectionspecs&gt; (optional, list of slobroks to use to find the pattern)<br>
 * config=&lt;comma-separated list of config servers&gt; (optional, list of config servers to use to find slobrok config)
 *
 * If both slobroks and config is specified, the list from slobroks is used.
 *
 * @author Haakon Humberset
 */
public class LoadBalancerPolicy extends SlobrokPolicy {
    private final String session;
    private final String pattern;
    private final LoadBalancer loadBalancer;

    LoadBalancerPolicy(String param) {
        this(parse(param));
    }

    private LoadBalancerPolicy(Map<String, String> params) {
        super();

        String cluster = params.get("cluster");
        session = params.get("session");

        if (cluster == null) {
            throw new IllegalArgumentException("Required parameter 'cluster' not set");
        }

        if (session == null) {
            throw new IllegalArgumentException("Required parameter 'session' not set");
        }

        pattern = cluster + "/*/" + session;
        loadBalancer = new AdaptiveLoadBalancer(cluster);
    }

    @Override
    public void select(RoutingContext context) {
        LoadBalancer.Node node = getRecipient(context);

        if (node != null) {
            context.setContext(node);
            Route route = new Route(context.getRoute());
            route.setHop(0, Hop.parse(node.entry.getSpecString() + "/" + session));
            context.addChild(route);
        } else {
            context.setError(ErrorCode.NO_ADDRESS_FOR_SERVICE, "Could not resolve any nodes to send to in pattern " + pattern);
        }
    }

    /**
       Finds the TCP address of the target.

       @return Returns a hop representing the TCP address of the target, or null if none could be found.
    */
    private LoadBalancer.Node getRecipient(RoutingContext context) {
        List<Mirror.Entry> lastLookup = lookup(context, pattern);
        return loadBalancer.getRecipient(lastLookup);
    }

    public void merge(RoutingContext context) {
        RoutingNodeIterator it = context.getChildIterator();
        Reply reply = it.removeReply();
        LoadBalancer.Node target = (LoadBalancer.Node)context.getContext();

        boolean busy = false;
        for (int i = 0; i < reply.getNumErrors(); i++) {
            if (reply.getError(i).getCode() == ErrorCode.SESSION_BUSY) {
                busy = true;
            }
        }
        loadBalancer.received(target, busy);

        context.setReply(reply);
    }

    @Override
    public void destroy() {

    }

    // For testing
    LoadBalancer getLoadBalancer() { return loadBalancer; }
}

File: documentapi/src/vespa/documentapi/messagebus/policies/loadbalancerpolicy.cpp
// Copyright 2017 Yahoo Holdings. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root.
#include "loadbalancerpolicy.h"
#include <vespa/messagebus/emptyreply.h>
#include <vespa/messagebus/errorcode.h>
#include <vespa/messagebus/error.h>
#include <vespa/messagebus/routing/ihopdirective.h>
#include <vespa/messagebus/routing/routingcontext.h>
#include <vespa/messagebus/routing/verbatimdirective.h>
#include <vespa/log/log.h>

LOG_SETUP(".loadbalancerpolicy");

namespace documentapi {

LoadBalancerPolicy::LoadBalancerPolicy(const string& param)
    : ExternSlobrokPolicy(parse(param))
{
    std::map<string, string> params(parse(param));

    if (params.find("cluster") != params.end()) {
        _cluster = params.find("cluster")->second;
    } else {
        _error = "Required parameter cluster not set";
        return;
    }

    if (params.find("session") != params.end()) {
        _session = params.find("session")->second;
    } else {
        _error = "Required parameter session not set";
        return;
    }

    _pattern = _cluster + "/*/" + _session;
    _loadBalancer.reset(new LoadBalancer(_cluster, _session));
}

void
LoadBalancerPolicy::doSelect(mbus::RoutingContext& context) {
    std::pair<string, int> node = getRecipient(context);

    if (node.second != -1) {
        context.setContext((uint64_t)node.second);
        mbus::Route route = context.getRoute();
        route.setHop(0, mbus::Hop::parse(node.first + "/" + _session));
        context.addChild(route);
    } else {
        context.setError(mbus::ErrorCode::NO_ADDRESS_FOR_SERVICE,
                             "Could not resolve any nodes to send to in pattern " + _pattern);
    }
}

void
LoadBalancerPolicy::merge(mbus::RoutingContext& context) {
    mbus::RoutingNodeIterator it = context.getChildIterator();
    mbus::Reply::UP reply = it.removeReply();

    uint64_t target = context.getContext().value.UINT64;

    bool busy = false;
    for (uint32_t i = 0; i < reply->getNumErrors(); i++) {
        if (reply->getError(i).getCode() == mbus::ErrorCode::SESSION_BUSY) {
            string lastSpec = _loadBalancer->getNodeInfo()[target].lastSpec;

            if (reply->getError(i).getMessage().find(lastSpec) == string::npos) {
                LOG(debug, "Received busy with message %s, doesn't contain target %s so not updating weight.", reply->getError(i).getMessage().c_str(), lastSpec.c_str());
            } else {
                LOG(debug, "Received busy for target node %d reducing weight of that node.", (int)target);
                busy = true;
            }
        }
    }

    _loadBalancer->received(target, busy);

    context.setReply(std::move(reply));
}

}

